File: frontend/src/app/features/patients/pesel-validator.ts
import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';

/**
 * Client-side mirror of the backend's PeselValidator (research.md #1) — UX feedback only, the
 * server is authoritative. Poland's standard 11-digit weighted checksum. An empty/absent value is
 * valid (PESEL is optional, FR-002).
 */
export function peselChecksumValidator(): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    const value = (control.value ?? '').toString().trim();
    if (value === '') {
      return null;
    }
    if (!/^\d{11}$/.test(value)) {
      return { peselChecksum: true };
    }
    const weights = [1, 3, 7, 9, 1, 3, 7, 9, 1, 3];
    const sum = weights.reduce((acc, w, i) => acc + w * Number(value[i]), 0);
    const expected = (10 - (sum % 10)) % 10;
    return expected === Number(value[10]) ? null : { peselChecksum: true };
  };
}

File: patient-service/src/main/java/com/dentalclinic/patient/record/PeselValidator.java
package com.dentalclinic.patient.record;

/**
 * Poland's standard 11-digit PESEL format + weighted-checksum validation (research.md #1, FR-002) —
 * server-side, authoritative (the frontend mirrors this for UX only). {@code null} is valid: PESEL
 * is optional (FR-002/FR-003), and a missing PESEL is not this validator's concern.
 */
public final class PeselValidator {

  private static final int[] WEIGHTS = {1, 3, 7, 9, 1, 3, 7, 9, 1, 3};

  private PeselValidator() {}

  public static boolean isValid(String pesel) {
    if (pesel == null) {
      return true;
    }
    if (pesel.length() != 11 || !pesel.chars().allMatch(Character::isDigit)) {
      return false;
    }

    int weightedSum = 0;
    for (int i = 0; i < WEIGHTS.length; i++) {
      weightedSum += WEIGHTS[i] * Character.digit(pesel.charAt(i), 10);
    }
    int expectedCheckDigit = (10 - (weightedSum % 10)) % 10;
    int actualCheckDigit = Character.digit(pesel.charAt(10), 10);

    return expectedCheckDigit == actualCheckDigit;
  }
}
